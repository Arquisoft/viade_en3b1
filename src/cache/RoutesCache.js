import { loadAllRoutes } from '../handler/RouteHandler';
import { loadMedia } from '../handler/MediaHandler';

export default {
    first: true,
    routes: [],
    addRouteToCache(route) {
        if (route && !this.routes.find((obj) => route.name === obj.name)) {
            this.routes.push(route);
        }
    },
    async getRoutes() {
        if (this.routes.length === 0 || this.first) {
            this.routes = await loadAllRoutes();
            this.first = false;
        }
        return this.routes;
    },
    async getRouteById(id) {
        if(this.routes.length === 0) {
            await this.getRoutes();
        }
        let selectedroute = this.routes.find((r) => r.getId() === id);
        if (selectedroute) {
            await selectedroute.calculateElevation();
            await new Promise((r) => setTimeout(r, 1000));
            return selectedroute;
        } else {
            return -1;
        }
        
        
    },
    clear() {
        this.routes = [];
        this.first = true;
    }
};