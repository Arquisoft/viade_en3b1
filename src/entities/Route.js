import { v4 as uuid } from "uuid";
import RouteElement from "./RouteElement";

class Route {

    /**
     * Constructor for a new Route. 
     * An id will be automatically generated (uuid v4).
     * TotalDistance will also be automatically generated.
     * 
     * RouteElements have name, latitude, longitude and elevation.
     * 
     * @param {String} name Name of the route.
     * @param {Date} date Date of the route.
     * @param {String} description Description of the route.
     * @param {Array} routeElements List of RouteElement objects, which are the trackpoints.
     * @param {Array} comments List of comments
     * @param {Array<String>} media List of urls of media (on the POD)
     */
    constructor(name, description, routeElements, comments, media, date) {
        this.id = uuid().toString();
        this.name = name;
        this.date = date;
        this.description = description;

        this.media = media;
        this.totalDistance = null;

        if (comments === null) {
            this.comments = [];
        } else {
            this.comments = comments;
        }

        if (routeElements[0] instanceof RouteElement) {
            this.routeElements = routeElements;
        } else {
            this.routeElements = generateRouteElements(routeElements);
        }
    }

    getId() {
        return this.id;
    }

    getName() {
        return this.name;
    }

    getDate() {
        return this.date;
    }

    getTime() {
        return this.time;
    }

    getDescription() {
        return this.description;
    }

    getRouteElements() {
        return this.routeElements;
    }

    getTotalDistance() {
        var total = 0;
        var aux = 0;

        for (let i = 1; i < this.routeElements.length; i++) {
            aux = this.routeElements[i].getLatitude() - this.routeElements[i - 1].getLatitude();
            total += this.addDistance(aux);
            aux = this.routeElements[i].getLongitude() - this.routeElements[i - 1].getLongitude();
            total += this.addDistance(aux);
            aux = this.routeElements[i].getElevation() - this.routeElements[i - 1].getElevation();
            total += this.addDistance(aux);
        }

        return total;
    }

    addDistance(aux) {
        if (aux < 0) {
            return -aux;
        } else {
            return aux;
        }
    }

    getComments() {
        return this.comments;
    }

    getMedia() {
        return this.media;
    }

    addMedia(media) {
        this.media.push(media);
    }

    addRouteElement(routeElement) {
        this.routeElements.push(routeElement);
    }

    getJsonLD() {
        let routePointsJson = [];
        this.routeElements.forEach((p) => routePointsJson.push(p.toJsonLatLng()));
        return JSON.stringify(
            {
                "@context": {
                    "@version": 1.1,
                    "comments": {
                        "@container": "@list",
                        "@id": "viade:comments"
                    },
                    "description": {
                        "@id": "schema:description",
                        "@type": "xs:string"
                    },
                    "media": {
                        "@container": "@list",
                        "@id": "viade:media"
                    },
                    "name": {
                        "@id": "schema:name",
                        "@type": "xs:string"
                    },
                    "points": {
                        "@container": "@list",
                        "@id": "viade:points"
                    },
                    "latitude": {
                        "@id": "schema:latitude",
                        "@type": "xs:double"
                    },
                    "longitude": {
                        "@id": "schema:longitude",
                        "@type": "xs:double"
                    },
                    "elevation": {
                        "@id": "schema:elevation",
                        "@type": "xsd:double"
                    },
                    "author": {
                        "@id": "schema:author",
                        "@type": "@id"
                    },
                    "rdf": "http://www.w3.org/1999/02/22-rdf-syntax-ns#",
                    "rdfs": "http://www.w3.org/2000/01/rdf-schema#",
                    "schema": "http://schema.org/",
                    "viade": "http://arquisoft.github.io/viadeSpec/",
                    "xsd": "http://www.w3.org/2001/XMLSchema#"
                },
                "name": this.name,
                "author": this.author,
                "description": this.description,
                "comments": this.comments,
                "media": this.media,
                "points": routePointsJson
            }
        );
    }

}

/**
 * Method that receives a list of points and returns
 * a list of RouteElements
 * @param {Array} points List of points containing name, latitude, longitude, elevation. 
 */
function generateRouteElements(points) {
    let routeElements = [];
    points.forEach((p) => routeElements.push(new RouteElement(p.lat, p.lng)));
    return routeElements;
}

export default Route;