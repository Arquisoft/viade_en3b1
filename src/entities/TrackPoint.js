class TrackPoint {

    /**
     * 
     * @param {Number} latitude 
     * @param {Number} longitude 
     * @param {String} name 
     * @param {Number} elevation 
     */
    constructor(latitude, longitude, name, elevation) {

        this.latitude = latitude;
        this.longitude = longitude;

        if (!name) {
            this.name = "";
        } else {
            this.name = name;
        }

        if(!elevation){
            this.elevation = "";
        } else {
            this.elevation = elevation;
        }
    }

    getName() {
        return this.name;
    }

    getLatitude() {
        return this.latitude;
    }

    getElevation() {
        return this.elevation;
    }

    getLongitude() {
        return this.longitude;
    }

    setElevation(num) {
        this.elevation = num;
    }

    toJson() {
        return {
            "latitude": this.latitude,
            "longitude": this.longitude,
            "elevation": this.elevation,
        };
    }
}

export default TrackPoint;