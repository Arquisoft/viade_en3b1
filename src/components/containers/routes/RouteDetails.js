import React from "react";
import NavBar from '../../graphic interface/NavBar.js';
import { useParams } from "react-router-dom";
import RouteDetailsCard from "../../graphic interface/RouteDetailsCard.js";
import cache from '../../../cache/RoutesChache';

const RouteDetails = () => {
  let { id } = useParams();

  let myroute=cache.getRoutesFromCache().filter((r) => (r.getId() === id));

  return (
    <div>
      <div>
        <NavBar />
      </div>
      <div>
        { <RouteDetailsCard route={myroute[0]} ></RouteDetailsCard> }
      </div>
    </div>
  );
};

export default RouteDetails;