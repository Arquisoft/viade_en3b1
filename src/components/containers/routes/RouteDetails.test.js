import RouteDetails from "./RouteDetails";
import Route from '../../../entities/Route';
import TrackPoint from '../../../entities/TrackPoint';
import RoutesCache from "../../../cache/RoutesCache";

test("Define RouteDetails", () => {
    expect(RouteDetails).toBeDefined();
});

test("Renders correctly", () => {
    const wrapper = render(<RouteDetails  match={{params: {id: 1}}} />);
    expect(wrapper).toMatchSnapshot();
});

test('Mounts correctly', () => {
    var route1 = new Route("Route1", "Description 1", [new TrackPoint(1,1), new TrackPoint(2,1)], null, null, null, "ID1");
    RoutesCache.addRouteToCache(route1);

    mount(<RouteDetails  match={{params: {id: 1}}} />);
});