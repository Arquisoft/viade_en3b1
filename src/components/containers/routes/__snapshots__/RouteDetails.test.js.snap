// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders correctly 1`] = `
<div
  style="text-align:center"
>
  <div
    class="MuiLinearProgress-root MuiLinearProgress-colorPrimary MuiLinearProgress-indeterminate"
    role="progressbar"
  >
    <div
      class="MuiLinearProgress-bar MuiLinearProgress-barColorPrimary MuiLinearProgress-bar1Indeterminate"
    />
    <div
      class="MuiLinearProgress-bar MuiLinearProgress-bar2Indeterminate MuiLinearProgress-barColorPrimary"
    />
  </div>
  <h5
    class="MuiTypography-root MuiTypography-h5"
  >
    Loading your route...
  </h5>
</div>
`;
